### ISSUE TITLE
getOrCreateIndex method

### ISSUE THREAD

**Issue by curquiza** [meilisearch/integration-guides#2]
We are thinking about creating a `getOrCreateIndex` method.

### Description

This method returns an `Index` object. It checks if the index already exists and if it's not, it creates the index.

### How

By trying to create the index, and catching the exception. In every situation, only one HTTP call is done.

An example in ruby:

```ruby
require 'meilisearch'

URL = 'http://127.0.0.1:7700'

def get_or_create_index(index_uid, primary_key = nil)
  client = MeiliSearch::Client.new(URL)

  begin
    index = client.create_index(uid: index_uid, primaryKey: primary_key)
  rescue MeiliSearch::HTTPError => e
    raise e unless e.message.include? 'index already exists'
    index = client.index(index_uid)
  end

  return index
end

index = get_or_create_index('indexUID')
puts index.show
```

(In ruby the `index` method is the equivalent of `getIndex` in other SDKs, and this method does not do any HTTP call)

This example is out of the MeiliSearch library. Of course, this method must be included in the client class.

### Tests

This method must be tested:
- Test with no index created before
- Test with an index already existing
- Test with a primary-key in parameter

### Steps

Do it in:
- [x] [JS](https://github.com/meilisearch/meilisearch-js/issues/343)
- [x] [Ruby](https://github.com/meilisearch/meilisearch-ruby/issues/29)
- [x] [Python](https://github.com/meilisearch/meilisearch-python/issues/70)
- [x] [PHP](https://github.com/meilisearch/meilisearch-php/issues/26)
- [x] [Swift](https://github.com/meilisearch/meilisearch-swift/issues/4)
- [x] .NET
- [x] Rust

Removed from the to-do list:
- [Go](https://github.com/meilisearch/meilisearch-go/issues/22) -> _not relevant since the client is divided into sub-clients that take an index themselves. The main client does not take any index._

### Feedback

Don't hesitate to suggest other ideas to implement the method, or if I forgot any tests 🙂

**Comment by alexisvisco** [comment]
In go it does not seem very relevant to me since the client is divided into sub-clients who take an index themselves. The main client does not take any index.

**Comment by curquiza** [comment]
You're right, I update the issue! Thanks

**Comment by ppamorim** [comment]
I really don't like the idea to compare the error message string. Would be nice to have a type of reason code or something.

**Comment by curquiza** [comment]
What do you suggest to not to check the message? 🙂
The `create_index` method could raise an exception if the index already exists, but it also means the method would have to check the error message to know if the error from MeiliSearch is about index existing...
What I mean it's if the weird check is not done in the `get_or_create_index` it would have done in `create_index`.

**Comment by shokme** [comment]
Meilisearch should have a way to check the existence of an index.
an endpoint like: 127.0.0.1:7700/indexes/{name}/exists

then, you just have to create a method `hasIndex(string $name): bool`

**Comment by eskombro** [comment]
I agree with @ppamorim about comparing error messages, seems really unreliable.
I like your idea @shokme, we could make an issue in the MeiliSearch repo to see if that is an option that pleases everyone.

Also, as for now, it seems that the possibilities of providing a method like this are:

1. Try to create the index, catch the error, check the error message to realize index already exists.
2. Try to request the index, catch the error, check the error message and realize the index doesn't exist yet.
3. List the existing indexes, look for your index inside the list to know if it exists.
4. ... ?

None of those options seems to be reliable or sustainable. The problem is that if we do not provide a method like this one, every SDK user is forced to check in it's own code every time if the index exists.

Can I request your POV @Kerollmops ?

**Comment by Kerollmops** [comment]
I prefer the one-call way proposed by @curquiza instead of the two-calls way proposed by @shokme as this is prone to races. A user can run multiple clients at the same time that which asks if the index exists or not, all the clients will see that the index doesn't exist and tries to create it at the same time. All the clients that were trying to create the index except the one that succeeded will return an error.

However, the error reporting story of MeiliSearch is not that good, we **must** introduce error codes as I already explained to @ppamorim. A unique code that helps clients to understand the reported error without having to do text comparison (which isn't a reliable solution). The documentation website will understand this unique code and redirect the user to the appropriate page.

In conclusion, I think that we should rely on string comparison as this is the safest and less race condition prone solution we have. We will improve the error reporting of the engine and you can track that in the related issue (I am searching for it right now...).

**Comment by bidoubiwa** [comment]
I'm building this method in the javascript library and I have a question about the primaryKey key.

The primary Key is set only once, either by inference either manually by the user. With this in mind, imagine the following scenario:

## Scenario 1: existing index without no primary key
I have an index `myIndex` that already exists, this index has no primaryKey.

Then, I call this method the following way.
```
getOrCreateIndex("myIndex", "id")
```
The primary key that is given here is useless because your method does not use this parameter unless the index does not exists.
This is not explicit, users could imagine that the primaryKey will be updated. 

## Scenario 2: existing index existing primary key
I have an index `myIndex` that already exists, this index has a primaryKey that is `id`.

Then, I call this method the following way.
```
getOrCreateIndex("myIndex", "newId")
```
The primary key that is given here is useless as the primary key already exists and can not be changed and for reason in scenario 1.

## Proposition
Taken the following parameters into account: 
- You only need an `uid` to create an index
- Options when creating will evolve and there will be more than just `primaryKey`
- because stating `primaryKey` as a parameter name in this function is not explicit in what will be done with it

I suggest creating a second parameter which would be `createOptions` in which for the moment there will only be `primaryKey`

a call in javascript would look like this
```javascript
await meili.getOrCreateIndex("uid", {
"primaryKey": "newId"
})
```

**Comment by curquiza** [comment]
I don't see the difference between your first scenario and second scenario example 😅 

But, I think (I'm not sure) I understood your suggestion in your conclusion.

You want as definition:

```ruby
def get_or_create_index(index_uid, options = nil)
```

and as usage:

```ruby
client.get_or_create_index('uid', {'primaryKey': 'id'})
```
Right?

I like this kind of idea, but I would improve your idea with this goal: **staying consistent with the `create_index` method.**
Look at [the usage in JS](https://github.com/meilisearch/meilisearch-js/#create-an-index-):

```js
// Create an index
const index = await client.createIndex({ uid: 'books' })
// Create an index and give the primary-key
const index = await client.createIndex({ uid: 'books', primaryKey: 'book_id' })
```
=> the `getOrCreateIndex` method should expect for an object with at least `uid`, as the `createIndex` method does.

For the ruby:
```ruby
# Create an index
client.create_index('books')
# Create an index and give the primary-key
client.create_index({uid: 'books', primaryKey: 'book_id'})
```

=> the `get_or_create_index` method should expect for a string or a hash table (= a dict  = an object) with at least the `uid` field.

For the PHP: same behavior than ruby.
For the Go: same behavior than the JS for the moment.

For the python, it's a little bit different... It does not have the same logic than the others 😞 
```python
# Create an index
client.create_index(uid='books')
# Create an index and give the primary-key
client.create_index(uid='books', primary_key='book_id')
```
> Options when creating will evolve and there will be more than just primaryKey

This is not possible here...

A solution is to break the python, but not too much, to propose the same behavior than you suggested in your comment:

```python
client.create_index(uid='indexUID', options={ primaryKey: 'id' })
client.get_or_create_index(uid='indexUID', options={ primaryKey: 'id' })

**Comment by bidoubiwa** [comment]
I really like your suggestion on staying consistent

```ruby
client.create_index(uid='indexUID', options={ primaryKey: 'id' })
client.get_or_create_index(uid='indexUID', options={ primaryKey: 'id' })
```
